File: scripts/scrape-remaining.mjs
import { readFile, writeFile, mkdir, access } from 'node:fs/promises'
import { createWriteStream } from 'node:fs'
import { get } from 'node:https'
import { chromium } from 'playwright'

const BASE = 'https://image2.fun'
const SLUGS_PATH = new URL('remaining_slugs.json', import.meta.url)
const OUTPUT_PATH = new URL('scraped_remaining.json', import.meta.url)

// Get remaining slugs
const html = await fetch(BASE).then(r => r.text())
const catUrls = ['', '?cat=portrait', '?cat=poster', '?cat=character', '?cat=ui-mockup',
  '?cat=infographic', '?cat=other', '?tag=ecommerce', '?tag=logo-branding', '?tag=food', '?tag=chinese']

const allSlugs = new Set()
for (const cat of catUrls) {
  const h = await fetch(BASE + '/' + cat).then(r => r.text())
  for (const m of h.matchAll(/href="(\/p\/[^"]+)"/g)) allSlugs.add(m[1])
}

// Read existing data for dedup
const existingText = await readFile(new URL('../src/data/templates.ts', import.meta.url), 'utf-8')
const existingUrls = new Set()
for (const m of existingText.matchAll(/"sourceUrl":\s*"([^"]+)"/g)) {
  const url = m[1]
  if (url.includes('/p/')) existingUrls.add('/p/' + url.split('/p/')[1])
}
const existingTitles = new Set([...existingText.matchAll(/"title":\s*"([^"]+)"/g)].map(m => m[1]))

const remaining = [...allSlugs].filter(s => !existingUrls.has(s))
console.log(`Remaining: ${remaining.length} cards`)

if (remaining.length === 0) { console.log('Nothing to scrape.'); process.exit(0) }

await writeFile(SLUGS_PATH, JSON.stringify(remaining))

// Scrape with Playwright
const results = []
const browser = await chromium.launch({ headless: true })
const context = await browser.newContext({ viewport: { width: 1440, height: 900 } })

for (let i = 0; i < remaining.length; i++) {
  const slug = remaining[i]
  process.stdout.write(`[${i+1}/${remaining.length}] ${slug} `)
  const page = await context.newPage()
  try {
    await page.goto(BASE + slug, { waitUntil: 'networkidle', timeout: 30000 })
    await new Promise(r => setTimeout(r, 800))

    const prompt = await page.evaluate(`() => {
      const el = document.querySelector('pre');
      if (el && el.textContent.length > 50) return el.textContent.trim();
      let best = '';
      for (const e of document.querySelectorAll('p, div')) {
        const t = (e.textContent || '').trim();
        if (t.length > best.length && t.length > 80) best = t;
      }
      return best || null;
    }`)
    const metaTitle = await page.evaluate('document.title')
    const ogImage = await page.evaluate(`() => {
      const m = document.querySelector('meta[property="og:image"]');
      return m ? m.content : null;
    }`)
    const title = ((metaTitle || '').split('·')[0] || slug.split('/').pop()).trim()

    results.push({ slug, title, image: ogImage, prompt })

    if (prompt && prompt.length > 50) console.log(`OK (${prompt.length}c)`)
    else console.log('NO PROMPT')
  } catch (e) {
    console.log(`ERROR: ${e.message.slice(0, 60)}`)
    results.push({ slug, title: '', image: null, prompt: null })
  }
  await page.close()
  await new Promise(r => setTimeout(r, 300))
}

await browser.close()
await writeFile(OUTPUT_PATH, JSON.stringify(results, null, 2))
console.log(`\nDone. ${results.filter(r => r.prompt && r.prompt.length > 50).length}/${results.length} with prompts`)

File: scripts/scrape_remaining.py
import json, time, os, re
from playwright.sync_api import sync_playwright

BASE = 'https://image2.fun'
OUTPUT = '/Users/wanghans/gpt-image2-templates/scripts/scraped_remaining.json'
IMG_DIR = '/Users/wanghans/gpt-image2-templates/public/images/'
os.makedirs(IMG_DIR, exist_ok=True)

# Get all slugs
cat_urls = ['', '?cat=portrait', '?cat=poster', '?cat=character', '?cat=ui-mockup',
            '?cat=infographic', '?cat=other', '?tag=ecommerce', '?tag=logo-branding', '?tag=food', '?tag=chinese']

import subprocess
all_slugs = set()
for cat in cat_urls:
    html = subprocess.check_output(['curl', '-sSL', f'{BASE}/{cat}'], text=True, timeout=15)
    slugs = set(re.findall(r'href="(/p/[^"]+)"', html))
    all_slugs.update(slugs)

# Read existing data for dedup
with open('/Users/wanghans/gpt-image2-templates/src/data/templates.ts') as f:
    text = f.read()
existing_urls = set()
for m in re.finditer(r'"sourceUrl":\s*"([^"]+)"', text):
    url = m.group(1)
    if '/p/' in url:
        existing_urls.add('/p/' + url.split('/p/')[1])
existing_titles = set(re.findall(r'"title":\s*"([^"]+)"', text))

remaining = sorted(set(all_slugs) - existing_urls)
print(f'Remaining: {len(remaining)} cards')

if not remaining:
    print('Nothing to scrape')
    exit()

# Save slugs reference
with open('/Users/wanghans/gpt-image2-templates/scripts/remaining_slugs.json', 'w') as f:
    json.dump(remaining, f)

# Scrape with Playwright
results = []
with sync_playwright() as p:
    browser = p.chromium.launch(headless=True)
    context = browser.new_context(viewport={"width": 1440, "height": 900})

    for i, slug in enumerate(remaining):
        print(f'[{i+1}/{len(remaining)}] {slug}', end=' ')
        page = None
        try:
            page = context.new_page()
            page.goto(f'{BASE}{slug}', wait_until='networkidle', timeout=30000)
            time.sleep(0.8)

            prompt = page.evaluate("""() => {
                const el = document.querySelector('pre');
                if (el && el.textContent.length > 50) return el.textContent.trim();
                let best = '';
                for (const e of document.querySelectorAll('p, div')) {
                    const t = (e.textContent || '').trim();
                    if (t.length > best.length && t.length > 80) best = t;
                }
                return best || null;
            }""")
            meta_title = page.evaluate('document.title')
            og_image = page.evaluate("""() => {
                const m = document.querySelector('meta[property="og:image"]');
                return m ? m.content : null;
            }""")
            t = (meta_title or '').split('·')[0].strip() or slug.split('/')[-1]

            results.append({'slug': slug, 'title': t, 'image': og_image, 'prompt': prompt})
            if prompt and len(prompt) > 50:
                print(f'OK ({len(prompt)}c)')
            else:
                print('NO PROMPT')
        except Exception as e:
            print(f'ERROR: {str(e)[:60]}')
            results.append({'slug': slug, 'title': '', 'image': None, 'prompt': None})
        finally:
            if page:
                page.close()
            time.sleep(0.3)

    browser.close()

with open(OUTPUT, 'w') as f:
    json.dump(results, f, ensure_ascii=False, indent=2)

with_prompt = sum(1 for r in results if r.get('prompt') and len(r['prompt']) > 50)
print(f'\nDone. {with_prompt}/{len(results)} with prompts')
